<h1>MỤC LỤC<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Giới-thiệu-đề-tài" data-toc-modified-id="Giới-thiệu-đề-tài-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Giới thiệu đề tài</a></span></li><li><span><a href="#Bài-toán-và-mô-tả-tập-dữ-liệu" data-toc-modified-id="Bài-toán-và-mô-tả-tập-dữ-liệu-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Bài toán và mô tả tập dữ liệu</a></span><ul class="toc-item"><li><span><a href="#Bài-toán" data-toc-modified-id="Bài-toán-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Bài toán</a></span></li><li><span><a href="#Mô-tả-tập-dữ-liệu" data-toc-modified-id="Mô-tả-tập-dữ-liệu-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Mô tả tập dữ liệu</a></span><ul class="toc-item"><li><span><a href="#Input" data-toc-modified-id="Input-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>Input</a></span></li><li><span><a href="#Output" data-toc-modified-id="Output-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>Output</a></span></li></ul></li></ul></li><li><span><a href="#Thuật-toán" data-toc-modified-id="Thuật-toán-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Thuật toán</a></span></li><li><span><a href="#Đánh-giá-mô-hình" data-toc-modified-id="Đánh-giá-mô-hình-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Đánh giá mô hình</a></span><ul class="toc-item"><li><span><a href="#Root-Mean-Square-Error-(RMSE):" data-toc-modified-id="Root-Mean-Square-Error-(RMSE):-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Root Mean Square Error (RMSE):</a></span></li></ul></li><li><span><a href="#Cài-đặt-hàm-và-thư-viện-cần-thiết" data-toc-modified-id="Cài-đặt-hàm-và-thư-viện-cần-thiết-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Cài đặt hàm và thư viện cần thiết</a></span></li><li><span><a href="#Chuẩn-bị-dữ-liệu-đầu-vào" data-toc-modified-id="Chuẩn-bị-dữ-liệu-đầu-vào-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Chuẩn bị dữ liệu đầu vào</a></span></li><li><span><a href="#Exploratory-data-analysis-(EDA)" data-toc-modified-id="Exploratory-data-analysis-(EDA)-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Exploratory data analysis (EDA)</a></span></li><li><span><a href="#Xây-dựng-mô-hình" data-toc-modified-id="Xây-dựng-mô-hình-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Xây dựng mô hình</a></span><ul class="toc-item"><li><span><a href="#Xây-dựng-mô-hình-Alternating-least-squares-chung" data-toc-modified-id="Xây-dựng-mô-hình-Alternating-least-squares-chung-8.1"><span class="toc-item-num">8.1&nbsp;&nbsp;</span>Xây dựng mô hình Alternating least squares chung</a></span><ul class="toc-item"><li><span><a href="#Xây-dựng-mô-hình" data-toc-modified-id="Xây-dựng-mô-hình-8.1.1"><span class="toc-item-num">8.1.1&nbsp;&nbsp;</span>Xây dựng mô hình</a></span></li><li><span><a href="#Đánh-giá-mô-hình" data-toc-modified-id="Đánh-giá-mô-hình-8.1.2"><span class="toc-item-num">8.1.2&nbsp;&nbsp;</span>Đánh giá mô hình</a></span></li></ul></li><li><span><a href="#Xây-dựng-và-đánh-giá-mô-hình-Alternating-least-squares-(tinh-chỉnh-các-siêu-tham-số-lần-1)" data-toc-modified-id="Xây-dựng-và-đánh-giá-mô-hình-Alternating-least-squares-(tinh-chỉnh-các-siêu-tham-số-lần-1)-8.2"><span class="toc-item-num">8.2&nbsp;&nbsp;</span>Xây dựng và đánh giá mô hình Alternating least squares (tinh chỉnh các siêu tham số lần 1)</a></span></li><li><span><a href="#Xây-dựng-và-đánh-giá-mô-hình-Alternating-least-squares-(tinh-chỉnh-các-siêu-tham-số-lần-2)" data-toc-modified-id="Xây-dựng-và-đánh-giá-mô-hình-Alternating-least-squares-(tinh-chỉnh-các-siêu-tham-số-lần-2)-8.3"><span class="toc-item-num">8.3&nbsp;&nbsp;</span>Xây dựng và đánh giá mô hình Alternating least squares (tinh chỉnh các siêu tham số lần 2)</a></span></li><li><span><a href="#Xây-dựng-và-đánh-giá-mô-hình-Alternating-least-squares-(tinh-chỉnh-các-siêu-tham-số-lần-3)" data-toc-modified-id="Xây-dựng-và-đánh-giá-mô-hình-Alternating-least-squares-(tinh-chỉnh-các-siêu-tham-số-lần-3)-8.4"><span class="toc-item-num">8.4&nbsp;&nbsp;</span>Xây dựng và đánh giá mô hình Alternating least squares (tinh chỉnh các siêu tham số lần 3)</a></span></li><li><span><a href="#Xây-dựng-và-đánh-giá-mô-hình-Alternating-least-squares-(tinh-chỉnh-các-siêu-tham-số-lần-4)" data-toc-modified-id="Xây-dựng-và-đánh-giá-mô-hình-Alternating-least-squares-(tinh-chỉnh-các-siêu-tham-số-lần-4)-8.5"><span class="toc-item-num">8.5&nbsp;&nbsp;</span>Xây dựng và đánh giá mô hình Alternating least squares (tinh chỉnh các siêu tham số lần 4)</a></span></li></ul></li><li><span><a href="#So-sánh-các-mô-hình" data-toc-modified-id="So-sánh-các-mô-hình-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>So sánh các mô hình</a></span></li><li><span><a href="#Triển-khai-mô-hình" data-toc-modified-id="Triển-khai-mô-hình-10"><span class="toc-item-num">10&nbsp;&nbsp;</span>Triển khai mô hình</a></span></li><li><span><a href="#Một-số-đề-xuất-cụ-thể-cho-người-dùng-và-sách" data-toc-modified-id="Một-số-đề-xuất-cụ-thể-cho-người-dùng-và-sách-11"><span class="toc-item-num">11&nbsp;&nbsp;</span>Một số đề xuất cụ thể cho người dùng và sách</a></span></li></ul></div>

# CHƯƠNG 1
# TỔNG QUAN VỀ ĐỀ TÀI

## Giới thiệu đề tài

Hệ thống gợi ý là một phương tiện có giá trị giúp giải quyết vấn đề tìm kiếm và chắt lọc thông tin. Mục đích cuối cùng mà hệ thống gợi ý hướng tới là “mang đến” cho người dùng những sản phẩm họ thích trong hàng ngàn sản phẩm cùng loại trong hệ thống, những sản phẩm mà người dùng chưa xem hoặc không xem nhưng lại đang liên quan đến những tác vụ tìm kiếm hiện tại của người dùng. Các sản phẩm gợi ý dựa trên số lượng sản phẩm đã bán, dựa trên thông tin cá nhân người dùng, dựa trên sự phân tích hành vi mua hàng trước đó của người sử dụng để đưa ra các dự đoán về hành vi mua hàng trong tương lai của người dùng đó. Nổi bật trong số đó là thuật toán Alternating Least Squares trong kỹ thuật phân rã ma trận

Thực tế có rất nhiều sản phẩm, dịch vụ cần được gợi ý đến nhu cầu của từng người dùng, đặc biệt là sách. Sách là một kho tàng tri thức khổng lồ với hàng trăm triệu cuốn sách khác nhau. Để tìm được các cuốn sách ưng ý và phù hợp với nhu cầu của mỗi người dùng không phải là chuyện dễ dàng. Vì thế nhóm chúng tôi đã lựa chọn đề tài “Recommendation system using Alternating least squares pyspark” để gợi ý sách.


## Bài toán và mô tả tập dữ liệu

### Bài toán

   Bài toán của hệ thống gợi ý sách đơn giản là sẽ gợi ý những cuốn sách tốt nhất mà người dùng có thể thích hoặc đang quan tâm từ hàng ngàn cuốn sách trong hệ thống. Trong đề tài này, chúng tôi sẽ tiến hành thực nghiệm việc áp dụng Spark Mlib để gợi ý sách.

### Mô tả tập dữ liệu

#### Input

Tập ratings.csv gồm các thông tin sau:
- book_id: Mã id của sách.
- user_id: Mã id của người đọc sách.
- rating: Xếp hạng (đánh giá) của người dùng cho sách.

Train và Test data: được split random theo tỉ lệ 80% - 20% từ ratings.csv.

Tập to_read.csv gồm các thông tin sau:
- book_id: Mã id của sách.
- user_id: Mã id của người đọc sách.

Bảng dữ liệu to_read.csv sẽ chứa thông tin về những cuốn sách mà người đọc đã chọn “Tôi sẽ đọc”.

Tập book.csv gồm các thông tin: id, book_id, best_book_id, work_id, books_count, isbn, isbn13, authors, original_publication_year, original_title, title, language_code, average_rating, ratings_count,work_ratings_count, ratings_1,ratings_2,ratings_3,ratings_4,ratings_5, work_text_reviews_count, image_url, small_image_url.

Vì có khá nhiều dữ liệu nhưng chúng tôi chỉ sài một vài dữ liệu trong book.csv như:
- book_id: Mã id của sách
- image_url: Đường dẫn hình ảnh bìa sách


#### Output

Đề xuất 10 tựa sách cho tất cả các user.

Để dữ liệu đầu ra được chính xác hơn, ta nối trái với dữ liệu vừa tìm được với bảng ratings để chọn ra những đầu sách mà người đọc chưa đọc.

Tiếp tục nối dữ liệu mới nhất với bảng to_read để dự đoán những đầu sách chính xác nhất.

Một số đề xuất cụ thể cho cả user và book trong hệ thống.

## Thuật toán

Để thực hiện bài toán, nhóm chúng tối dùng thuật toán Alternating least squares sử dụng thư viện spark.

Alternating least squares (ALS) hay còn gọi là phương pháp bình phương tối thiểu thay phiên nhau. Trong hàm mất mát của ALS 2 biến tối ưu là Q và P. Bởi hàm mất mát có 2 biến nên khiến hàm không lồi (khó tìm cực tiểu). Vì vậy ý tưởng của thuật toán ALS là thay phiên cố định 1 biến rồi tối ưu hàm theo biến còn lại. Khi cố định 1 biến hàm sẽ trở thành hàm bậc 2 và dễ dàng tính toán bằng phương pháp bình phương tối thiểu. Ta thực hiện lần lượt việc cố định 1 biến rồi tính toán lại biến còn lại cho đến khi các điểm hội tụ lại tại điểm cực tiểu.

\begin{equation}
\min _{q_{i}, p_{u}} \sum_{(u, i) \in K}\left(r_{u} i-q_{i}^{T} p_{u}\right)^{2}+\lambda\left(\left\|q_{i}\right\|^{2}+\left\|p_{u}\right\|^{2}\right)
\end{equation}

Trong đó:
- $q_{i}$ đánh giá những yếu tố tiềm ẩn mà item i sở hữu.
- $p_{u}$ đánh giá những yếu tố tiềm ẩn mà user u sở hữu.

## Đánh giá mô hình

### Root Mean Square Error (RMSE):

Lỗi trung bình bình phương gốc (RMSE) là thước đo mức độ hiệu quả của mô hình của bạn. Nó thực hiện điều này bằng cách đo sự khác biệt giữa các giá trị dự đoán và giá trị thực tế . R-MSE càng nhỏ tức là sai số càng bé thì mức độ ước lượng cho thấy độ tin cậy của mô hình có thể đạt cao nhất.

\begin{equation}
R M S E=\sqrt{\sum_{i=1}^{n} \frac{\left(\hat{y}_{i}-y_{i}\right)^{2}}{n}}
\end{equation}

Với:
- $\hat{y}_{i}$: là giá trị dự đoán
- $y_{i}$ là giá trị thực tế
- n là tổng lượng quan sát

Giá trị bình phương trung bình lỗi lấy căn thường xuyên được sử dụng trong đánh giá độ khớp của mô hình so với dữ liệu huấn luyện. Được tính bằng căn bậc hai của giá trị tuyệt đối của hệ số tương quan giữa giá trị thực và giá trị dự đoán.

# CHƯƠNG 2

# XÂY DựNG MÔ HÌNH

## Cài đặt hàm và thư viện cần thiết

In [1]:
import os
import pandas as pd
import numpy as np

# PySpark
from pyspark import SparkContext
from pyspark.sql import SparkSession

# Exploratory Data Analysis (EDA)
from pyspark.sql.functions import col, min, max, avg, lit

# Machine Learning (ML)
from pyspark.ml.recommendation import ALS # Alternating Least Squares model
from pyspark.ml.evaluation import RegressionEvaluator # Performance metric

# Visualization
import seaborn as sns
import matplotlib.pyplot as plt

from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

pd.set_option('display.max_columns', 200)
pd.set_option('display.max_colwidth', 400)

from matplotlib import rcParams
sns.set(context='notebook', style='whitegrid', rc={'figure.figsize': (18,4)})
rcParams['figure.figsize'] = 18,4

%matplotlib inline
%config InlineBackend.figure_format = 'retina'

# Setting random seed for reproducability
SEED = 23
np.random.seed = SEED
np.random.set_state = SEED

In [2]:
sc = SparkContext(appName = "Book-Recommendation")
print(sc)

<SparkContext master=local[*] appName=Book-Recommendation>


In [3]:
spark = SparkSession.Builder().getOrCreate()
print(spark)

## Chuẩn bị dữ liệu đầu vào

Chúng tôi tiến hành load 2 tập dữ liệu là ratings.csv và to_read.csv vào dataframe (Thay đổi đường dẫn tùy vào máy mỗi người sử dụng).

In [4]:
# Đọc file csv vào dataframe
ratings = spark.read.csv('D:\MLFINAL/ratings.csv',
                         header = True, #sử dụng dòng đầu tiên làm tiêu đề
                         inferSchema=True) #tự suy ra các loại dữ liệu
print(type(ratings))

<class 'pyspark.sql.dataframe.DataFrame'>


In [5]:
to_read = spark.read.csv('D:\MLFINAL/to_read.csv',
                         header = True,
                         inferSchema=True)
print(type(to_read))

<class 'pyspark.sql.dataframe.DataFrame'>


Kiểm tra thuộc tính của các cột trong dataframe và kiểm tra các cột nào có giá trị bị thiếu (null) trên tập ratings.

In [6]:
ratings.printSchema()

root
 |-- book_id: integer (nullable = true)
 |-- user_id: integer (nullable = true)
 |-- rating: integer (nullable = true)



Chúng tôi chuyển đổi thuộc tính cột rating từ integer về double cho dễ dàng xử lý dữ liệu.

In [7]:
ratings = ratings.select(ratings.user_id,
                         ratings.book_id,
                         ratings.rating.cast("double"))
ratings.printSchema()

root
 |-- user_id: integer (nullable = true)
 |-- book_id: integer (nullable = true)
 |-- rating: double (nullable = true)



## Exploratory data analysis (EDA)

Chúng tôi in ra 10 dòng dữ liệu để kiểm tra những dữ liệu đã có trong tập ratings.csv

In [8]:
#in ra 10 dòng trong tập dữ liệu ratings
print(ratings.show(n = 10))

+-------+-------+------+
|user_id|book_id|rating|
+-------+-------+------+
|    314|      1|   5.0|
|    439|      1|   3.0|
|    588|      1|   5.0|
|   1169|      1|   4.0|
|   1185|      1|   4.0|
|   2077|      1|   4.0|
|   2487|      1|   4.0|
|   2900|      1|   5.0|
|   3662|      1|   4.0|
|   3922|      1|   5.0|
+-------+-------+------+
only showing top 10 rows

None


Chúng tôi sẽ tính toán tỉ lệ các ô trống (Sparsity) trong ma trận.

In [9]:
# Đếm tổng số xếp hạng trong tập dữ liệu ratings
numerator = ratings.select("rating").count()

# Đếm các id (book và user) riêng biệt
num_users = ratings.select("user_id").distinct().count()
num_items = ratings.select("book_id").distinct().count()

# Tính toán mẫu số
denominator = num_users * num_items

# Tính toán tỉ lệ thưa thớt trong ma trận
sparsity = (1.0 - (numerator * 1.0)/ denominator) * 100
print("The ratings dataframe is ", "%.2f" % sparsity + "% empty.")

The ratings dataframe is  99.82% empty.


Nhìn vào kết quả ta có thể nhận thấy, hầu như ma trận hoàn toàn trống rỗng (99.82%).

Tiếp theo, ta tiến hành nhóm các dữ liệu user_id và đếm số lượng rating (làm tương tự với book_id)

In [10]:
# Nhóm user_id, đếm rating
(ratings.groupBy("user_id")
    .count()
    .filter("`count` > 1")
    .sort(col("count").desc())
    .show(n = 10))

+-------+-----+
|user_id|count|
+-------+-----+
|  12874|  200|
|  30944|  200|
|  12381|  199|
|  28158|  199|
|  52036|  199|
|  45554|  197|
|   6630|  197|
|  37834|  197|
|  15604|  196|
|   9806|  196|
+-------+-----+
only showing top 10 rows



In [11]:
# Nhóm book_id, đếm rating
(ratings.groupBy("book_id")
    .count()
    .filter("`count` > 1")
    .sort(col("count").desc())
    .show(n = 10))

+-------+-----+
|book_id|count|
+-------+-----+
|    148|  100|
|    463|  100|
|    471|  100|
|    496|  100|
|    833|  100|
|   1088|  100|
|   1238|  100|
|   1342|  100|
|   1580|  100|
|   1591|  100|
+-------+-----+
only showing top 10 rows



Tóm tắt các thống kê của tập dữ liệu ratings để chúng ta có cái nhìn trực quan hơn về tập dữ liệu.

In [12]:
# Số lượng xếp hạng nhỏ nhất cho item
print("Item with the fewest ratings: ")
ratings.groupBy("book_id").count().select(min("count")).show()

Item with the fewest ratings: 
+----------+
|min(count)|
+----------+
|         8|
+----------+



In [13]:
# Xếp hạng trung bình trên mỗi item
print("Avg num ratings per item: ")
ratings.groupBy("book_id").count().select(avg("count")).show()

Avg num ratings per item: 
+----------+
|avg(count)|
+----------+
|   98.1756|
+----------+



In [14]:
# Số lượng xếp hạng nhỏ nhất cho user
print("User with the fewest ratings: ")
ratings.groupBy("user_id").count().select(min("count")).show()

User with the fewest ratings: 
+----------+
|min(count)|
+----------+
|         2|
+----------+



In [15]:
# Xếp hạng trung bình trên mỗi user
print("Avg num ratings per user: ")
ratings.groupBy("user_id").count().select(avg("count")).show()

Avg num ratings per user: 
+------------------+
|        avg(count)|
+------------------+
|18.376684636118597|
+------------------+



## Xây dựng mô hình

Chúng tôi tiến hành chia tập ratings.csv thành hai tập train và test trước khi vào xây dựng mô hình.
- Tập train: 80%
- Tập test: 20%

In [16]:
# Chia ratings thành 2 tập train và test  với tỉ lệ 80% và 20%.
train, test = ratings.randomSplit([.8, .2])

### Xây dựng mô hình Alternating least squares chung

#### Xây dựng mô hình

Chúng tôi xây dựng mô hình Alternating least squares (không có các siêu tham số) từ tập train và kiểm tra kết quả mô hình ở tập test.

In [17]:
# Xây dựng mô hình ALS, không có các siêu tham số
als = ALS(userCol="user_id", itemCol="book_id", ratingCol="rating", 
          nonnegative = True, # Phân tích ma trận thành nhân tử không âm
          coldStartStrategy = "drop", # Xóa đi những người dùng mới không xuất hiện trên train và test
          implicitPrefs = False) # Hủy những dữ liệu không rõ ràng
als_model = als.fit(train)

#Áp dụng mô hình ở tập test để kiểm tra
test_pred = als_model.transform(test)

In [18]:
# Xác nhận mô hình als đã được tạo
type(als)

pyspark.ml.recommendation.ALS

Sau khi áp dụng mô hình lên tập test, ta xem kết quả các dự đoán (n=10)

In [19]:
# Xem kết quả các dự đoán 
test_pred.show(n = 10)

+-------+-------+------+----------+
|user_id|book_id|rating|prediction|
+-------+-------+------+----------+
|  28767|    148|   3.0| 3.0026605|
|  24326|    148|   5.0| 3.4562576|
|  18313|    148|   4.0| 3.2681909|
|   5461|    148|   4.0| 3.6424634|
|  13407|    148|   4.0| 3.8904939|
|  50999|    148|   3.0| 3.3447177|
|  22534|    148|   4.0|  3.395548|
|   5885|    148|   4.0| 3.3181143|
|   9246|    148|   3.0| 2.7905517|
|  21658|    148|   4.0|    3.2999|
+-------+-------+------+----------+
only showing top 10 rows



Hình trên mô tả id của user và book, rating là chỉ số xếp hạng thực tế của người dùng cho cuốn sách đó. Prediction là nhãn mà mô hình dự đoán.
Ví dụ: người dùng có id là 28767 xếp hạng cho sách có id là 148 với xếp hạng thực tế là 3.0. Mô hình dự đoán xếp hạng của người dùng cho sách là 3.0026605. Có thể nhận thấy xếp hạng thực tế và dự đoán khá trùng khớp (Có thể thay đổi khi mọi người chạy lại lệnh này vì sẽ lấy random 10 người nên kết quả có thể bị thay đổi).

#### Đánh giá mô hình

Đánh giá mô hình bằng cách tính giá trị Root-mean-square error (RMSE). Giá trị này là độ chênh lệch giữa giá trị dự đoán và thực tế. Giá trị càng nhỏ thì mô hình càng hiệu quả.

In [20]:
# Tính RMSE trên test_pred
evaluator = RegressionEvaluator(metricName = "rmse", 
                                labelCol = "rating", 
                                predictionCol = "prediction")
rmse = evaluator.evaluate(test_pred)
print("Root-mean-square error = " + str(rmse))

Root-mean-square error = 0.9185619748126058


RMSE trả về giá trị khoảng 0.91. Điều này cho thấy mô hình khá tốt, nhưng đây chỉ là mô hình chung của ALS, chưa có các siêu tham số, chúng tôi sẽ tiếp tục xây dựng thêm các mô hình có siêu tham số và tinh chỉnh các siêu tham số sao cho mô hình được tốt nhất.

### Xây dựng và đánh giá mô hình Alternating least squares (tinh chỉnh các siêu tham số lần 1)

Để xây dựng mô hình Alternating least squares với các siêu tham số, ta cần quan tâm đến 3 tham số:
- maxIter: là số lần lặp tối đa để chạy (mặc định là 10).
- regParam chỉ định tham số chính quy hóa trong ALS (mặc định là 1.0).
- rank: là số lượng các yếu tố tiềm ẩn trong mô hình (mặc định là 10).

In [27]:
# Xây dựng mô hình ALS, tinh chỉnh các siêu tham số lần 1
als = ALS(maxIter=15, regParam=0.1, rank=10, userCol="user_id", itemCol="book_id", ratingCol="rating")
model = als.fit(train)
predictions = model.transform(test)
new_predictions = predictions.filter(col('prediction') != np.nan)
evaluator = RegressionEvaluator(metricName="rmse", labelCol="rating", predictionCol="prediction")
rmse = evaluator.evaluate(new_predictions)
print("Root-mean-square error = " + str(rmse))

Root-mean-square error = 0.9005038113166655


Chúng tôi điều chỉnh lần lượt maxIter=15, regParam=0.1, rank=10 và nhận được RMSE là khoảng 0.9. RMSE đã giảm so với mô hình ALS chung nhưng vẫn chưa tốt nhất.

### Xây dựng và đánh giá mô hình Alternating least squares (tinh chỉnh các siêu tham số lần 2)

In [37]:
# Xây dựng mô hình ALS, tinh chỉnh các siêu tham số lần 2
als = ALS(maxIter=10, regParam=0.2, rank=8, userCol="user_id", itemCol="book_id", ratingCol="rating")
model = als.fit(train)
predictions = model.transform(test)
new_predictions = predictions.filter(col('prediction') != np.nan)
evaluator = RegressionEvaluator(metricName="rmse", labelCol="rating", predictionCol="prediction")
rmse = evaluator.evaluate(new_predictions)
print("Root-mean-square error = " + str(rmse))

Root-mean-square error = 0.8750651630605232


Với maxIter=10, regParam=0.2, rank=8, ta nhận được RMSE khoảng 0.875.

### Xây dựng và đánh giá mô hình Alternating least squares (tinh chỉnh các siêu tham số lần 3)

In [38]:
# Xây dựng mô hình ALS, tinh chỉnh các siêu tham số lần 3
als = ALS(maxIter=10, regParam=0.2, rank=5, userCol="user_id", itemCol="book_id", ratingCol="rating")
model = als.fit(train)
predictions = model.transform(test)
new_predictions = predictions.filter(col('prediction') != np.nan)
evaluator = RegressionEvaluator(metricName="rmse", labelCol="rating", predictionCol="prediction")
rmse = evaluator.evaluate(new_predictions)
print("Root-mean-square error = " + str(rmse))

Root-mean-square error = 0.8733854939189009


Với maxIter=10, regParam=0.2, rank=5, ta nhận được RMSE khoảng 0.873.

### Xây dựng và đánh giá mô hình Alternating least squares (tinh chỉnh các siêu tham số lần 4)

In [39]:
# Xây dựng mô hình ALS, tinh chỉnh các siêu tham số lần 4
final_als = ALS(maxIter=10, regParam=0.2, rank=4, userCol="user_id", itemCol="book_id", ratingCol="rating")
final_model = final_als.fit(train)
predictions = final_model.transform(test)
new_predictions = predictions.filter(col('prediction') != np.nan)
evaluator = RegressionEvaluator(metricName="rmse", labelCol="rating", predictionCol="prediction")
rmse = evaluator.evaluate(new_predictions)
print("Root-mean-square error = " + str(rmse))

Root-mean-square error = 0.8725672296920765


Với maxIter=10, regParam=0.2, rank=4, ta nhận được RMSE khoảng 0.872.

## So sánh các mô hình

 Mô hình:                                        RMSE

Alternating least squares chung:                 0.9186

Alternating least squares (tinh chỉnh lần 1):    0.9005

Alternating least squares (tinh chỉnh lần 2):    0.8751

Alternating least squares (tinh chỉnh lần 3):    0.8734

Alternating least squares (tinh chỉnh lần 4):    0.8726


Nhận xét: Từ các giá trị RMSE đã tính được từ 5 mô hình ALS trên, ta nhận thấy mô hình ALS (tinh chỉnh lần 4) là tốt nhất.

## Triển khai mô hình

Sau khi đã tìm được mô hình tốt nhất, chúng tôi sẽ sử dụng mô hình tốt nhất để gợi ý sách cho người dùng kết hợp với tập to_read.csv.

Chúng tôi sẽ tạo n đề xuất cho tất cả người dùng (với n=10). Trong đây cũng có thể đề xuất những cuốn sách mà người dùng đã đọc. Chúng tôi sẽ lọc ra ở các bước tiếp theo.

In [40]:
# Tạo n đề xuất cho tất cả người dùng
ALS_recommendations = final_model.recommendForAllUsers(numItems = 10) # n = 10
ALS_recommendations.show(n = 10)

+-------+--------------------+
|user_id|     recommendations|
+-------+--------------------+
|    148|[[9842, 4.4344187...|
|    463|[[9531, 4.4544363...|
|    471|[[9842, 3.6001418...|
|    496|[[6902, 4.779125]...|
|    833|[[9842, 4.4773254...|
|   1088|[[6590, 4.4429903...|
|   1238|[[9531, 5.088133]...|
|   1342|[[4868, 4.4561067...|
|   1580|[[6590, 4.8356867...|
|   1591|[[4868, 3.4241772...|
+-------+--------------------+
only showing top 10 rows



Chuyển đổi đầu ra thành user_id, book_id, prediction cho dễ dàng quan sát.

In [44]:
# Chuyển đổi đầu ra thành user_id, book_id, prediction
ALS_recommendations.registerTempTable("ALS_recs_temp")

clean_recs = spark.sql("""SELECT user_id,
                            bookIds_and_ratings.book_id AS book_id,
                            bookIds_and_ratings.rating AS prediction
                        FROM ALS_recs_temp
                        LATERAL VIEW explode(recommendations) exploded_table
                            AS bookIds_and_ratings""")
clean_recs.show()

+-------+-------+----------+
|user_id|book_id|prediction|
+-------+-------+----------+
|    148|   9842| 4.4344187|
|    148|   5207| 4.3260894|
|    148|   4868|  4.310271|
|    148|   6920|  4.300663|
|    148|   3491| 4.2974644|
|    148|   9076|  4.295326|
|    148|   6590|  4.263487|
|    148|   9566|  4.263332|
|    148|   3885| 4.2625923|
|    148|   4483|  4.256458|
|    463|   9531| 4.4544363|
|    463|   8109|  4.412524|
|    463|   6920|  4.376691|
|    463|   5580| 4.3218427|
|    463|   6590| 4.3171372|
|    463|   7254|  4.307837|
|    463|   5207|  4.289475|
|    463|   8415| 4.2812295|
|    463|   9566|  4.280412|
|    463|   1788|   4.23633|
+-------+-------+----------+
only showing top 20 rows



Tiếp theo chúng tôi sẽ tìm những cuốn sách chưa được đọc và đếm xem có bao nhiêu cuốn.

In [45]:
# Đề xuất những cuốn sách chưa đọc
(clean_recs.join(ratings, ["user_id", "book_id"], "left")
    .filter(ratings.rating.isNull()).show())

+-------+-------+----------+------+
|user_id|book_id|prediction|rating|
+-------+-------+----------+------+
|     20|   7063|  4.861883|  null|
|     31|   1788| 3.7420979|  null|
|     31|   6590| 3.7460775|  null|
|     35|    862| 3.6461272|  null|
|     53|   8580| 3.3309677|  null|
|     54|   3628|  4.191738|  null|
|     55|   7910| 3.8336086|  null|
|     78|   8253|   2.43172|  null|
|    103|   6590| 3.8811398|  null|
|    132|   6590| 5.2160845|  null|
|    152|   6590|  4.878297|  null|
|    200|   7537| 4.2685437|  null|
|    264|   9123|  4.264239|  null|
|    280|   5207|  4.589646|  null|
|    281|   9067| 3.7224245|  null|
|    283|   4721| 3.9074273|  null|
|    325|    862| 4.3510556|  null|
|    326|   7593| 4.3583555|  null|
|    329|   5207| 4.4639506|  null|
|    334|   6902| 2.9111981|  null|
+-------+-------+----------+------+
only showing top 20 rows



In [46]:
# Đếm xem tổng cộng có bao nhiêu sách chưa có ratings
new_books = (clean_recs.join(ratings, ["user_id", "book_id"], "left")
    .filter(ratings.rating.isNull()))
print(new_books.count())

526957


Sau đó chúng tôi nối bảng những cuốn sách chưa có ratings (nhiều users khác nhau đánh giá) với tập to_read.csv. Bảng này lưu trữ những thông tin về các cuốn sách mà người dùng đã đánh dấu “Tôi muốn đọc”, nghĩa là người dùng có khả năng thích và đánh giá cao cho cuốn sách đó.

In [48]:
# Join những cuốn sách chưa có ratings với tập to_read.csv
recommendations = new_books.join(to_read.withColumn('to_read', lit(1)), 
                              on = ["user_id", "book_id"], 
                              how = "left")
print(recommendations.show())

+-------+-------+----------+------+-------+
|user_id|book_id|prediction|rating|to_read|
+-------+-------+----------+------+-------+
|     20|   7063|  4.861883|  null|   null|
|     31|   1788| 3.7420979|  null|   null|
|     31|   6590| 3.7460775|  null|   null|
|     35|    862| 3.6461272|  null|   null|
|     53|   8580| 3.3309677|  null|   null|
|     54|   3628|  4.191738|  null|   null|
|     55|   7910| 3.8336086|  null|   null|
|     78|   8253|   2.43172|  null|   null|
|    103|   6590| 3.8811398|  null|   null|
|    132|   6590| 5.2160845|  null|   null|
|    152|   6590|  4.878297|  null|   null|
|    200|   7537| 4.2685437|  null|   null|
|    264|   9123|  4.264239|  null|   null|
|    280|   5207|  4.589646|  null|   null|
|    281|   9067| 3.7224245|  null|   null|
|    283|   4721| 3.9074273|  null|   null|
|    325|    862| 4.3510556|  null|   null|
|    326|   7593| 4.3583555|  null|   null|
|    329|   5207| 4.4639506|  null|   null|
|    334|   6902| 2.9111981|  nu

Kiểm tra xem là có bao nhiêu cuốn sách nằm trong tập to_read.csv

In [49]:
(recommendations
     .groupby('to_read')
     .count()
     .show())

+-------+------+
|to_read| count|
+-------+------+
|   null|525777|
|      1|  1180|
+-------+------+



Ta nhận thấy chỉ có 1180 cuốn sách được người dùng đánh dấu là “Tôi muốn đọc” , ta sẽ đánh giá những cuốn sách này. Giả thuyết đặt ra là trung bình xếp hạng sẽ cao, vì người dùng đã muốn đọc từ trước.

Ta xử lý đầu ra dưới dạng bảng với các cột: user_id, book_id, prediction, to_read thể hiện giá trị dự đoán của một người dùng lên cuốn sách đã đánh dấu là “Tôi muốn đọc”.

In [50]:
(recommendations
     .filter(recommendations.to_read.isNotNull())
     .select(['user_id', 'book_id', 'prediction','to_read'])
     .show())

+-------+-------+----------+-------+
|user_id|book_id|prediction|to_read|
+-------+-------+----------+-------+
|  29212|    562|   4.90353|      1|
|  33454|   1568|  4.677004|      1|
|  41891|   3628| 5.0178556|      1|
|  18149|   7275| 3.9924896|      1|
|  30947|   1338|  5.319868|      1|
|  34876|   9842|  4.886486|      1|
|  42271|    862| 5.0453444|      1|
|  52844|   2236| 3.8458447|      1|
|   1394|   3885| 4.5630035|      1|
|  23857|   7593|  4.983957|      1|
|  24025|   8521| 3.4669182|      1|
|  36806|   3628|  4.951229|      1|
|  43590|   9570| 3.6903286|      1|
|  51037|   1308| 3.9853904|      1|
|  18516|    192|  3.932479|      1|
|  41832|    862|  4.664405|      1|
|  45483|   6902|  4.163578|      1|
|  16062|   9294| 3.9472816|      1|
|  16212|   9569|  4.548699|      1|
|  19783|   6963| 3.6460238|      1|
+-------+-------+----------+-------+
only showing top 20 rows



In [51]:
to_read_recs = (recommendations
     .filter(recommendations.to_read.isNotNull())
     .select(['user_id', 'book_id', 'prediction','to_read']))
to_read_recs.show(n = 5)

+-------+-------+----------+-------+
|user_id|book_id|prediction|to_read|
+-------+-------+----------+-------+
|  29212|    562|   4.90353|      1|
|  33454|   1568|  4.677004|      1|
|  41891|   3628| 5.0178556|      1|
|  18149|   7275| 3.9924896|      1|
|  30947|   1338|  5.319868|      1|
+-------+-------+----------+-------+
only showing top 5 rows



Cuối cùng ta sẽ in ra xem thang điểm đánh giá 1180 cuốn sách trong to_read.csv. Hầu hết các cuốn sách đều nằm trong khoảng đánh giá từ 3-4 sao, cho thấy người dùng đều thích các cuốn sách mình đá đánh dấu trước đó. Qua đó cho thấy mô hình hoạt động tốt.

In [52]:
(to_read_recs
     .withColumn('pred_trunc', to_read_recs.prediction.substr(1,1))
     .groupby('pred_trunc')
     .count()
     .sort('pred_trunc')
    .show())

+----------+-----+
|pred_trunc|count|
+----------+-----+
|         1|    6|
|         2|   30|
|         3|  251|
|         4|  695|
|         5|  194|
|         6|    4|
+----------+-----+



## Một số đề xuất cụ thể cho người dùng và sách

Đầu tiên chúng tôi load tập dữ liệu book.csv vào dataframe.

In [54]:
books = spark.read.csv('D:\MLFINAL/books.csv',
                         header = True,
                         inferSchema=True)
books.printSchema()

root
 |-- id: integer (nullable = true)
 |-- book_id: integer (nullable = true)
 |-- best_book_id: integer (nullable = true)
 |-- work_id: integer (nullable = true)
 |-- books_count: integer (nullable = true)
 |-- isbn: string (nullable = true)
 |-- isbn13: double (nullable = true)
 |-- authors: string (nullable = true)
 |-- original_publication_year: double (nullable = true)
 |-- original_title: string (nullable = true)
 |-- title: string (nullable = true)
 |-- language_code: string (nullable = true)
 |-- average_rating: string (nullable = true)
 |-- ratings_count: string (nullable = true)
 |-- work_ratings_count: string (nullable = true)
 |-- work_text_reviews_count: string (nullable = true)
 |-- ratings_1: double (nullable = true)
 |-- ratings_2: integer (nullable = true)
 |-- ratings_3: integer (nullable = true)
 |-- ratings_4: integer (nullable = true)
 |-- ratings_5: integer (nullable = true)
 |-- image_url: string (nullable = true)
 |-- small_image_url: string (nullable = true)


Sử dụng lại mô hình ALS tốt nhất

In [55]:
# Xây dựng mô hình ALS, tinh chỉnh các siêu tham số lần 4
final_als = ALS(maxIter=10, regParam=0.2, rank=4, userCol="user_id", itemCol="book_id", ratingCol="rating")
final_model = final_als.fit(train)
final_predictions = final_model.transform(test)
final_new_predictions = final_predictions.filter(col('prediction') != np.nan)
final_evaluator = RegressionEvaluator(metricName="rmse", labelCol="rating", predictionCol="prediction")
rmse = final_evaluator.evaluate(final_new_predictions)
print("Root-mean-square error = " + str(rmse))

Root-mean-square error = 0.8725672296920765


Kết hợp với books.csv để view ra id người đọc, tên cuốn sách và số điểm dự đoán.

In [56]:
final_predictions.join(books, "book_id").select("user_id","title","prediction").show(5)

+-------+----------+----------+
|user_id|     title|prediction|
+-------+----------+----------+
|   6213|Lysistrata|  2.816645|
|  27745|Lysistrata| 3.6269598|
|  11326|Lysistrata|  4.085428|
|  12466|Lysistrata|  3.644628|
|  52425|Lysistrata| 3.1227784|
+-------+----------+----------+
only showing top 5 rows



Tạo đề xuất cho một người dùng cụ thể, ở đây chúng tôi chọn user_id=35982.

In [57]:
for_one_user = predictions.filter(col("user_id")==35982).join(books_df, "book_id").select("user_id","title","image_url","prediction")
for_one_user.count()
for_one_user.show()

6

+-------+--------------------+--------------------+----------+
|user_id|               title|           image_url|prediction|
+-------+--------------------+--------------------+----------+
|  35982|Another Bullshit ...|https://s.gr-asse...| 3.7899556|
|  35982|The Door Into Summer|https://images.gr...| 3.1902225|
|  35982|City of Glass (Th...|https://images.gr...| 3.7808537|
|  35982|       Metamorphoses|https://s.gr-asse...|  3.898744|
|  35982|           The Lover|https://images.gr...| 3.4095907|
|  35982|The Lord of the R...|https://s.gr-asse...| 3.8616562|
+-------+--------------------+--------------------+----------+



In [59]:
from IPython.display import Image
from IPython.display import display

In ra bìa các cuốn sách đề xuất.

In [61]:
for book in for_one_user.take(10):
    print(book.title)
    display(Image(url=book.image_url))

Another Bullshit Night in Suck City


The Door Into Summer


City of Glass (The New York Trilogy, #1)


Metamorphoses


The Lover


The Lord of the Rings: Weapons and Warfare


Tạo 5 đề xuất sách hàng đầu cho mỗi người dùng và 5 đề xuất người dùng hàng đầu cho mỗi cuốn sách.

In [62]:
#Tạo 5 đề xuất sách hàng đầu cho mỗi người dùng
userRecomments = final_model.recommendForAllUsers(5)
#Tạo 5 đề xuất người dùng hàng đầu cho mỗi cuốn sách
bookRecomments = final_model.recommendForAllItems(5)

In [63]:
userRecomments.select("user_id","recommendations.book_id").show(10,False)

+-------+------------------------------+
|user_id|book_id                       |
+-------+------------------------------+
|148    |[9842, 5207, 4868, 6920, 3491]|
|463    |[9531, 8109, 6920, 5580, 6590]|
|471    |[9842, 4778, 2636, 8521, 8926]|
|496    |[6902, 8854, 9531, 562, 2236] |
|833    |[9842, 9531, 9569, 1308, 7275]|
|1088   |[6590, 5207, 9566, 1788, 6920]|
|1238   |[9531, 6902, 1308, 6963, 8548]|
|1342   |[4868, 8187, 7844, 862, 3628] |
|1580   |[6590, 5207, 6920, 1788, 862] |
|1591   |[4868, 8926, 8854, 3885, 3491]|
+-------+------------------------------+
only showing top 10 rows



In [64]:
bookRecomments.select("book_id","recommendations.user_id").show(10,False)

+-------+-----------------------------------+
|book_id|user_id                            |
+-------+-----------------------------------+
|1580   |[48681, 43252, 14211, 9448, 46064] |
|4900   |[38723, 15449, 9448, 33923, 49360] |
|5300   |[28724, 38723, 21791, 19698, 16505]|
|6620   |[43252, 48681, 14211, 21347, 33143]|
|7240   |[38723, 9448, 46796, 49360, 51626] |
|7340   |[43252, 48681, 38723, 9448, 14211] |
|7880   |[38723, 48681, 43252, 9448, 48274] |
|9900   |[48681, 43252, 30440, 14211, 42106]|
|471    |[48681, 43252, 38723, 9448, 48274] |
|1591   |[48681, 43252, 14211, 9448, 46064] |
+-------+-----------------------------------+
only showing top 10 rows



Tạo 10 đề xuất hàng đầu cho người dùng cụ thể và 10 đề xuất hàng đầu cho cuốn sách cụ thể.

In [67]:
#Tạo 10 đề xuất hàng đầu cho người dùng cụ thể
users = ratings.select("user_id").distinct().limit(5);
userSubsetRecs = final_model.recommendForUserSubset(users, 10)
userSubsetRecs.select("user_id","recommendations.book_id").show(10,False)

+-------+------------------------------------------------------------+
|user_id|book_id                                                     |
+-------+------------------------------------------------------------+
|32592  |[6920, 6902, 5207, 9076, 4868, 9531, 6590, 862, 4483, 3753] |
|35982  |[1788, 6590, 3628, 5207, 9566, 5580, 6920, 7254, 7844, 7593]|
|19984  |[6902, 9531, 1308, 6920, 9076, 5207, 3753, 4868, 8926, 8854]|
|1088   |[6590, 5207, 9566, 1788, 6920, 7254, 3628, 5580, 8109, 9076]|
|3918   |[6590, 5207, 9566, 1788, 7254, 6920, 3628, 5580, 8109, 9076]|
+-------+------------------------------------------------------------+



In [70]:
#Tạo 10 đề xuất hàng đầu cho cuốn sách cụ thể
books = ratings.select("book_id").distinct().limit(5);
bookSubsetRecs = final_model.recommendForItemSubset(books, 10)
bookSubsetRecs.select("book_id","recommendations.user_id").show(10,False)

+-------+---------------------------------------------------------------------+
|book_id|user_id                                                              |
+-------+---------------------------------------------------------------------+
|471    |[48681, 43252, 38723, 9448, 48274, 27079, 43544, 33923, 38866, 14211]|
|463    |[38723, 15449, 9448, 33923, 11193, 51739, 48274, 13033, 7172, 50307] |
|833    |[38723, 43252, 48681, 9448, 15449, 33923, 48274, 11193, 7172, 43544] |
|496    |[48681, 43252, 9448, 38723, 33923, 48274, 46796, 15449, 27079, 17128]|
|148    |[43252, 48681, 38723, 9448, 48274, 27079, 46796, 33923, 15449, 38866]|
+-------+---------------------------------------------------------------------+

